packages/lib/server/checkBookingLimits.ts

```diff
@@ -0,0 +1,89 @@
+import dayjs from "@calcom/dayjs";
+import prisma from "@calcom/prisma";
+import { BookingLimit } from "@calcom/types/Calendar";
+
+import { HttpError } from "../http-error";
+
+export async function checkBookingLimits(
+  bookingLimits: BookingLimit,
+  eventStartDate: Date,
+  eventId: number,
+  returnBusyTimes?: boolean
+) {
+  const limitCalculations = Object.entries(bookingLimits).map(
+    async ([key, limitingNumber]) =>
+      await checkLimit({ key, limitingNumber, eventStartDate, eventId, returnBusyTimes })
+  );
+  await Promise.all(limitCalculations)
+    .then((res) => {
+      if (returnBusyTimes) {
+        return res;
+      }
+    })
+    .catch((error) => {
+      throw new HttpError({ message: error.message, statusCode: 401 });
+    });
+  return true;
+}
+
+export async function checkLimit({
+  eventStartDate,
+  eventId,
+  key,
+  limitingNumber,
+  returnBusyTimes = false,
+}: {
+  eventStartDate: Date;
+  eventId: number;
+  key: string;
+  limitingNumber: number;
+  returnBusyTimes?: boolean;
+}) {
+  {
+    const limitKey = key as keyof BookingLimit;
+    // Take PER_DAY and turn it into day and PER_WEEK into week etc.
+    const filter = limitKey.split("_")[1].toLocaleLowerCase() as "day" | "week" | "month" | "year"; // Have to cast here
+    const startDate = dayjs(eventStartDate).startOf(filter).toDate();
+    // this is parsed above with parseBookingLimit so we know it's safe.
+
+    const endDate = dayjs(startDate).endOf(filter).toDate();
+
+    // This allows us to easily add it within dayjs
+    const bookingsInPeriod = await prisma.booking.count({
+      where: {
+        status: "ACCEPTED",
+        eventType: {
+          AND: [
+            {
+              id: eventId,
+              bookings: {
+                some: {
+                  startTime: {
+                    gte: startDate,
+                  },
+                  endTime: {
+                    lte: endDate,
+                  },
+                },
+              },
+            },
+          ],
+        },
+      },
+    });
+    if (bookingsInPeriod >= limitingNumber) {
+      // This is used when getting availbility
+      if (returnBusyTimes) {
+        return {
+          start: startDate,
+          end: endDate,
+        };
+      }
+
+      throw new HttpError({
+        message: `Booking limit of ${limitingNumber}/${filter} reached for this eventType`,
+        statusCode: 401,
```

Comment:
Is 401 the right statuscode here?

Reply:
Good shout - should be 403 (forbidden) for if we wanna go super fancy use 422 (unprocurable entity )

Reply:
Setting to 403 for consistency